src/rules/tictactoepoker/tictactoepoker.ts

```diff
@@ -0,0 +1,752 @@
+    public get_rank_value(): number {
+        assert(this.is_normal());
+        if (this.value == Card.AceValue) {
+            return 14;
+        }
+        return this.value;
+    }
+}
+
+type Deck = Card[];
+
+// Returns a shuffled standard deck with special cards.
+function new_deck(): Deck {
+    // Normal cards
+    let result = [];
+    for (let suit in Suit) {
+        for (let value = 2; value <= 13; value++) {
+            result.push(Card.create_normal(suit, value));
+        }
+        result.push(Card.create_normal(suit, Card.AceValue));
+    }
+    result.push(Card.create_special(CardType.Joker));
+    result.push(Card.create_special(CardType.Joker));
+    result.push(Card.create_special(CardType.Thief));
+    shuffle(result);
+    return result;
+}
+
+// Combination types. The values should be such that a higher value represents a better combination.
+enum CombinationType {
+    None = 0,
+    Pair = 1,
+    ThreeOfAKind = 2,
+    Straight = 3,
+    Flush = 4,
+    StraightFlush = 5,
+}
+
+// Score type.
+interface ScoreType {
+    combination_type: CombinationType,
+    rank_card?: Card,
+}
+
+// Returns the score value for the given score type.
+function compute_score_value(score_type: ScoreType): number {
+    let rank_value = 0;
+    if (score_type.rank_card) {
+        rank_value = score_type.rank_card.get_rank_value();
+    }
+
+    switch(score_type.combination_type) {
+        case CombinationType.StraightFlush:
+            return 1000 + 30 * rank_value;
+        case CombinationType.Flush:
+            return 750 + 15 * rank_value;
+        case CombinationType.Straight:
+            return 500 + 10 * rank_value;
+        case CombinationType.ThreeOfAKind:
+            return 250 + 5 * rank_value;
+        case CombinationType.Pair:
+            return 100 + rank_value;
+        case CombinationType.none:
+            return 0;
+    }
+
+    throw("should not get here");
+}
+
+// Represents an explanation for how a single scoring line (row, column, or diagonal)
+export class ScoreExplanation {
+    private readonly score_type: ScoreType
+    private readonly where: string
+    private readonly score_value: number;
+
+    constructor(score_type: ScoreType, where: string) {
+        this.score_type = score_type;
+        this.where = where;
+        this.score_value = compute_score_value(score_type);
+    }
+
+    public get_score_value(): number {
+        return this.score_value;
+    }
+
+    public toString(): string {
+        return CombinationType[this.score_type.combination_type] + " in " + this.where + " " +
+            "with ranking card " + this.score_type.rank_card.toString() + " worth " +
+            this.score_value + " points";
+    }
+}
+
+export interface ScoreWithExplanations {
+    score: number,
+    explanations: ScoreExplanation[]
+}
+
+// Return score type for the given three cards (some of all of which may be empty cards)
+function compute_score(a: Card, b: Card, c: Card): ScoreType {
```

Comment:
don't have unit tests just for this function ah

Reply:
No, this is tested indirectly via the tests on scoring on Board.